source/lac/trilinos_tpetra_vector.cc

```diff
@@ -0,0 +1,686 @@
+    {
+      // Check that casting will work.
+      Assert(dynamic_cast<const Vector<Number> *>(&V) != nullptr,
+             ExcVectorTypeNotCompatible());
+
+      // Downcast V. If fails, throws an exception.
+      const Vector<Number> &down_V = dynamic_cast<const Vector<Number> &>(V);
+
+      reinit(down_V.locally_owned_elements(),
+             down_V.get_mpi_communicator(),
+             omit_zeroing_entries);
+    }
+
+
+
+    template <typename Number>
+    Vector<Number> &
+    Vector<Number>::operator=(const Vector<Number> &V)
+    {
+      // Distinguish three cases:
+      //  - First case: both vectors have the same layout.
+      //  - Second case: both vectors have the same size but different layout.
+      //  - Third case: the vectors have different size.
+      if (vector->getMap()->isSameAs(*(V.trilinos_vector().getMap())))
+        *vector = V.trilinos_vector();
+      else
+        {
+          if (size() == V.size())
+            {
+              Tpetra::Import<int, types::global_dof_index> data_exchange(
+                vector->getMap(), V.trilinos_vector().getMap());
+
+              vector->doImport(V.trilinos_vector(),
+                               data_exchange,
+                               Tpetra::REPLACE);
+            }
+          else
+            vector = std_cxx14::make_unique<
+              Tpetra::Vector<Number, int, types::global_dof_index>>(
+              V.trilinos_vector());
+        }
+
+      return *this;
+    }
+
+
+
+    template <typename Number>
+    Vector<Number> &
+    Vector<Number>::operator=(const Number s)
+    {
+      Assert(s == Number(0.),
+             ExcMessage("Only 0 can be assigned to a vector."));
+
+      vector->putScalar(s);
+
+      return *this;
+    }
+
+
+
+    template <typename Number>
+    void
+    Vector<Number>::import(
+      const ReadWriteVector<Number> &                 V,
+      VectorOperation::values                         operation,
+      std::shared_ptr<const CommunicationPatternBase> communication_pattern)
+    {
+      // If no communication pattern is given, create one. Otherwsie, use the
+      // one given.
+      if (communication_pattern == nullptr)
+        {
+          // The first time import is called, a communication pattern is
+          // created. Check if the communication pattern already exists and if
+          // it can be reused.
+          if ((source_stored_elements.size() !=
+               V.get_stored_elements().size()) ||
+              (source_stored_elements != V.get_stored_elements()))
+            {
+              const Teuchos::MpiComm<int> *mpi_comm =
+                dynamic_cast<const Teuchos::MpiComm<int> *>(
+                  vector->getMap()->getComm().get());
+              Assert(mpi_comm != nullptr, ExcInternalError());
+              create_tpetra_comm_pattern(V.get_stored_elements(),
+                                         *(mpi_comm->getRawMpiComm())());
```

Comment:
Is the communication pattern stored somewhere?

Reply:
Yes, it's stored in the member variable 
``` 
std::shared_ptr<const TpetraWrappers::CommunicationPattern> tpetra_comm_pattern;
```